### ISSUE TITLE
Don't run the pre/post scripts

### ISSUE THREAD

**Issue by zkochan** [pnpm/pnpm#2891]
When running `pnpm run foo`, don't run the `prefoo` script and the `postfoo` script.

Yarn 2 already deprecated the pre/post scripts: 

> In particular, we intentionally don't support arbitrary pre and post hooks for user-defined scripts (such as prestart). This behavior, inherited from npm, caused scripts to be implicit rather than explicit, obfuscating the execution flow. It also led to surprising executions with yarn serve also running yarn preserve.

https://yarnpkg.com/advanced/lifecycle-scripts

**Comment by aminya** [comment]
Does this affect `postinstall` as well? Because many packages use `postinstall` and not running that is not a good approach!

**Comment by zkochan** [comment]
no, the standard lifecycle scripts like postinstall or prepublishOnly are not affected.

**Comment by Djaler** [comment]
Is there are a way to enable it again? I think this is useful feature in some cases

**Comment by zkochan** [comment]
I don't see much value in it. You may just add `"start": "pnpm prestart && node server.js"` if you need it.

It is not some action that you do frequently. Mostly you just add a script once and never change it.

**Comment by Djaler** [comment]
Makes sense

**Comment by TigersWay** [comment]
> I don't see much value in it. You may just add `"start": "pnpm prestart && node server.js"` if you need it.
> 
> It is not some action that you do frequently. Mostly you just add a script once and never change it.

May I disagree? 
Nearly all my scripts have been written with these "pre" & "post" hooks to deal specifically with env variables or clean actions. And I know I am not the only one.
Just to be clear I honestly don't care what yarn is doing!

**Comment by Heartnett** [comment]
I agree with @TigersWay! I also don't want a yarn clone.

**Comment by cdaringe** [comment]
if you need pre/post, you can get similar ergonomics + explicitness (which this changes aims to acheive) with https://github.com/mysticatea/npm-run-all  (`run-p`/`run-s`)

**Comment by Heartnett** [comment]
> if you need pre/post, you can get similar ergonomics + explicitness (which this changes aims to acheive) with https://github.com/mysticatea/npm-run-all (`run-p`/`run-s`)

@cdaringe does npm-run-all even work with pnpm?

**Comment by TigersWay** [comment]
> @cdaringe does npm-run-all even work with pnpm?

npm-run-all does work, yes.
But pre/post hooks somehow allowed "multiline" scripts. And there, npm-run-all is no use!

**Comment by davidenke** [comment]
What a mess, this behavior is different to `npm` and (at least for me) unexpected.
It would make sense to have at least a flag (like `--run-scripts`) to restore the default behavior.

**Comment by jmcdo29** [comment]
When working with Typescript projects, `prebuild` is a very common script I work with to `rimraf dist` and while I can add that to the build script it would be nice to allow the `prebuild` to work directly. I'd be happy to try and figure out how we can make this work via a config or something so we can add it back in.

**Comment by cdaringe** [comment]
> What a mess ...

@davidenke, wrong attitude. you realize this software is offered to you for free? this unhelpful feedback is not welcome.

@jmcdo29, here's a demo of a setup that could work for you. 

```json
{
  "scripts": {
    "compile": "tsc index.ts",
    "serve": "node index.js",
    "start": "run-s compile serve"
  },
  "devDependencies": {
    "npm-run-all": "^4.1.5",
    "typescript": "^4.2.4"
  }
}
```

to see it in action, you can copy and paste the following command right into your terminal and run it! (assumes you are on a mac or *nix system)


```bash
export DIR=/tmp/pnpm-scripts-demo \
 && rm -rf $DIR && mkdir -p $DIR \
 && cd $DIR \
 && echo "console.log(1+2)" > index.ts \
 && echo '{"scripts":{"compile":"tsc index.ts","serve":"node index.js","start":"run-s compile serve"},"devDependencies":{"npm-run-all":"^4.1.5","typescript":"^4.2.4"}}' > package.json \
 && pnpm install \
 && pnpm start
```

```txt
$ pnpm start

@ start /private/tmp/pnpm-scripts-demo
> run-s compile serve


> @ compile /private/tmp/pnpm-scripts-demo
> tsc index.ts


> @ serve /private/tmp/pnpm-scripts-demo
> node index.js

3
```

**Comment by jmcdo29** [comment]
I understand it could work, was just going to not have to add another dev dependency and work with yet another package. But if that's what I have to do then so be it.

**Comment by jmcdo29** [comment]
For what it's worth, this might be something good to document on the [pnpm run](https://pnpm.io/cli/run) page, as, up until v6 to my knowledge, pnpm mirrored npm's functionality, including running pre/post scripts of custom commands. As [npm v7 still supports this](https://docs.npmjs.com/cli/v7/using-npm/scripts#pre--post-scripts), it's definitely a bit divergent from what I thought the functionality would be. I know the mention is in the changelog, but it would be nice to have it documented in the docs page as well, to help make it clearer. Just my two cents.

**Comment by TigersWay** [comment]
@cdaringe I'm sorry to share @jmcdo29's idea about the "mess" which is for me still a somehow polite word. We are here part of this community and any of us can be surprised with such a change.
Even so, "npm-run-all" is not exactly s solution: it's not maintained anymore and have some pretty bad bugs which are going to be a pain for this "unneeded" but forced transition.

**Comment by cdaringe** [comment]
> idea about the "mess" which is for me still a somehow polite word.

In other words, "someone did something I don't agree with, so I can be rude."  Insults and negativity--especially over something offered for free--are never productive. Consider sending patches, proposing solutions, debating pros and cons, etc.

Complaining like this rude, and helps no one. Most of us are taught this from a young age, as it has nearly global acceptance, across cultures. For reasons not yet clear to me, software engineers, in issues trackers that are not their own, feel comfortable waiving this value in favor selfishness.

Consider productive & helpful questions like:

- "<authors>, is it a design goal of yours to have parity between pnpm & npm? If so, do you see this as a violation?"
- "<authors>, do you agree or disagree that lifecycle hooks are helpful? is removing them more hurtful than helpful?"
- "<authors>, is the yarn teams' analysis and decision justifiable, or could they be over-reacting to to edge cases?"

**Comment by TigersWay** [comment]
@cdaringe 
- Again, I believe that in most area in this world, "mess" is not a rude word.
- Free does not mean "accept and shup-up".
- Before trying to give advice, have a look around to see if (maybe) all what you believe should have been done, has not been done already.
- This issue means simply that some dev are going to re-think their logic, their tools, their hopes.
And these was of course my last words here.

**Comment by jmcdo29** [comment]
@cdaringe maybe I'm missing something here. Yes there was a complaint about the functionality, but it wasn't just absurdly rude. Perhaps "mess" means something different to you than how I'm reading it. Either way, the dev did mention why they thought this (it's different than npm and unexpected) and mentioned a possible solution (using a `--run-scripts` flag to possibly allow for running these scripts).

You're right that in open source, complaining never gets anywhere, trust me, as a maintainer of NestJS and a few other packages I get that. It's one of the reasons I mentioned at the very least this should be well documented.

[pnpm stands for performant npm](https://pnpm.io/faq#what-does-pnpm-stand-for) which is why so many of us are confused at this different functionality. If this is a design goal of pnpm, to be like npm, then this definitely would be something that is no longer aligned.

I see where the yarn team is coming from, so maybe instead of just `pre` and `post` we could use something like `pre:` and `post:` to help avoid `preserve` from being ran. 

As I mentioned in my [first comment on this thread](https://github.com/pnpm/pnpm/issues/2891#issuecomment-817387883) I do find great use in these lifecycle scripts, and have been chasing why I've been having CI failures recently until I found that these scripts are no longer running. Now, it's my fault for using `@latest` of the pnpm GitHub Action, but it was still a shock to see to say the least.

**Comment by zkochan** [comment]
I was not personally offended.

We for sure need to document it prominently.

> If this is a design goal of pnpm, to be like npm, then this definitely would be something that is no longer aligned.

To some extend. pnpm was named before Yarn existed😀 We try to pick the good ideas from both npm and from Yarn. And sometimes have our own way.

This change aligns with our philosophy of being "strict" and explicit.

We can come up with some features to make it easier. My personal favorite would be to bundle something like `npm-run-all` with pnpm (which we were already asked for). But I am a bit afraid to include too much functionality into pnpm. I'd personally prefer to recommend some alternative script runner for advanced usages. This pre/post doesn't seem too much of an issue for now (and this is also the feedback that I recieved from Yarn's maintainer).

**Comment by jmcdo29** [comment]
> was not personally offended

Good to hear. I don't think there was an intent to offend here.

> We for sure need to document it prominently

I can make a PR for this if you'd like. Something along the lines of

> Pnpm does not run pre/post scripts for custom scripts. This leads to more explicit scripts and less vague execution flow.

Or we can just take what yarn said and drop it into the run page.

> our philosophy of being strict and explicit

Definitely a fair reasoning. Thanks for taking some time to answer more on this thread.

Would you say that it's pnpm's advice to use something like `npm-run-all`? It seems that these scripts won't be re-enabled anytime soon, which is a shame, but understandable.

**Comment by zkochan** [comment]
> It seems that these scripts won't be re-enabled anytime soon, which is a shame, but understandable.

It depends on the community. For now, it seems like it is not a huge deal.

> I can make a PR for this if you'd like. Something along the lines of

You can edit this page I guess https://pnpm.io/cli/run

> Would you say that it's pnpm's advice to use something like npm-run-all

`npm-run-all` is good for things like running commands in parallel. Sure it can also run commands in sequence but is it better then just writing something like `"test": "pnpm pretest && jest"`?

With npm-run-all it would be an overkill:

```json
"scripts": {
  "jest": "jest",
  "pretest": "<some command>.",
  "test": "run-s pretest jest"
}
```

**Comment by shellscape** [comment]
@zkochan unfortunately this change effects several very large monorepos for us. We have several deployment and smoke/e2e test scripts that run during github workflows which depend on pre/post scripts running, which kick off recursive scripts. the setup is complicated but elegant and took a long time to setup.

We're in desperate need of an option to allow the legacy behavior.

**Comment by zkochan** [comment]
https://github.com/pnpm/pnpm/pull/3348

**Comment by zkochan** [comment]
v6.1.0 has the `enable-pre-post-scripts` setting and it is added to the docs: https://pnpm.io/cli/run

**Comment by beenotung** [comment]
Just my view. I use pnpm to install packages, and run npm scripts with npm run.
If you would prefer the npm style's run script, you can still use npm run alone with pnpm-installed project.

(I have set `nr` as alias to `npm run`, so "'pnpm script-name' is shorter than 'npm run script-name'" isn't a reason to use 'pnpm script-name'. Did I missed any benefit to run script with pnpm/yarn?

**Comment by shellscape** [comment]
@beenotung point it moot now that 6.1.0 has the new option.

Thanks again @zkochan

**Comment by mwittmann** [comment]
I just ran into this and was glad to see that an `enable-pre-post-scripts` configuration setting has been added for ongoing compatibility with `npm` behavior, even if it's no longer pnpm's default behavior. 

IMO, keeping pnpm be highly compatible with npm&mdash;more or less a drop-in replacement &mdash;is very attractive and facilitates adoption of pnpm. That was certainly why I felt comfortable moving to pnpm.

You might consider updating the [pnpm vs npm](https://pnpm.io/pnpm-vs-npm) page to mention this difference in default behavior.

**Comment by beenotung** [comment]
I also prefer to have pnpm behavior the same as npm. Some tools like ionic and create-react-app doesn't support creating project using pnpm, so I added a script called npm, which call pnpm (and added into PATH).

But I've to check for edge cases to callback npm due to the inconsistent behaviors. A bit pity but still worth to do so.

**Comment by Camo30** [comment]
@zkochan: For me it is unclear how to set the `enable-pre-post-scripts` setting. Do you have an example?

**Comment by zkochan** [comment]
Configuration settings are set via the `.npmrc` file

**Comment by mwittmann** [comment]
> @zkochan: For me it is unclear how to set the `enable-pre-post-scripts` setting. Do you have an example?

Or command line: `pnpm config set enable-pre-post-scripts true`

**Comment by beenotung** [comment]
For better compatibility (when other developers may run your npm scripts without enabling pre/post scripts). I am updating existing project to adopt npm-run-all.

Additional benefit with this approach is it becomes easier to run a certain npm script when you're debugging, while ensuring the complete workflow is executed when running via an 'entry npm script' during CI/CD for example.

**Comment by stalniy** [comment]
oh guys... You are making such a big problem out of nothing. This hooks worked for years and nobody cared. Don't want to use them - don't use them. Strategy like "put everything into single line with &&`" works but makes readability harder.

Previously, I had this:
```
{
    "prerelease": "npm run lint && npm test && NODE_ENV=production npm run build",
    "release": "dx semantic-release"
}
```
and now I need to do that:
```
{
   "release": "npm run lint && npm test && NODE_ENV=production npm run build && dx semantic-release"
}
```
So, now there is no clear understanding of what is the main command to doing. This is especially hard when we have post hooks and add them at the end. Unfortunately

So, what that script is actually doing is not obvious anymore!  I think this change is ridiculous, and `enable-pre-post-scripts` must be `true` by default.

I could rewrite this to make it cleaner:

```
{
   "prerelease": "npm run lint && npm test && NODE_ENV=production npm run build",
   "release": "npm run prerelese && dx semantic-release && npm run postrelease"
   "postrelease": "some commands"
}
```

But.... What's a point, what are we fixing by this?

**Comment by gluons** [comment]
@stalniy For now, you can solve this by adding this config to your project.

**.npmrc**:
```ini
enable-pre-post-scripts=true
```

**Comment by zkochan** [comment]
> But.... What's a point, what are we fixing by this?

it is right in the issue

> In particular, we intentionally don't support arbitrary pre and post hooks for user-defined scripts (such as prestart). This behavior, inherited from npm, caused scripts to be implicit rather than explicit, obfuscating the execution flow. It also led to surprising executions with yarn serve also running yarn preserve.

**Comment by stalniy** [comment]
So, what’s bad in that yarn serve calls yarn preserve? And why for custom hooks explicit is good but for standard implicit hooks are still supported?

I see this as an abstraction. Hooks are a good pattern that used not only by yarn or npm. Webpack compiler provides hooks to intercept compilation. Rollup plug-ins also have hooks. And there are millions examples like that. Shall they also be more explicit?

It looks like I don’t see the point. And it’s never expressed in clear way. Too me this change is still weird, there are a lot of packages and apps that use custom hooks by intention.

**Comment by zkochan** [comment]
> what’s bad in that yarn serve calls yarn preserve

"preserve" has its own meaning in the English language, so running it before "serve" doesn't make sense.

> why for custom hooks explicit is good but for standard implicit hooks are still supported?

Standard hooks are not implicit. There are standard scripts that run on some events, like "postinstall". But prepostinstall, postpostinstall will not be executed. So the same rules apply.

**Comment by stalniy** [comment]
ok, could this be solved with another prefix, like `pre_serve` (or `pre:serve` or `pre.serve`)? the feature is very useful, I think its removal doesn't help real users.

There are few points which make me confused:
1. Removal of this hooks doesn't improve codebase. I looked through the PR, just few lines were removed. So, maintenance of the project is not affected in any way
2. Removal of this hooks force a lot of people to change a lot of packages (thankfully pnpm has a config option but yarn doesn't!). 
3. Removal of this hooks makes my scripts to be harder to reason about

There are 2  very big pain points. And the question is whose this problem? Who will feel this pain? The answer - developers, open source contributors. I'm as a one of those guys, need to go over my **14 open source node projects** and rewrite all scripts in package.json (that have been working for years without changes!), make them more uglier and unclear for no good reason (because in English there is "preserve" word which may conflict with "preserve" hook). I will spend this time doing stupid thing which will not help me to progress, improve my libs, provide better support, contribute to open source and help companies and people around to solve real problems.

As I said thankfully, I don't use yarn and pnpm has an option. But this is a road in the wrong direction. 

The real solution to this issue is to change hook names and use some extra symbol to separate hook name and script name. Even though it may conflict with existing hooks in some codebase but impact will be much slower in comparison to complete removal of this feature. There is a big difference in renaming hook vs merging scripts together and ensure that you put all symbols in place.

**P.S.**: Sorry for criticizing this task, I like pnpm very much, you guys do amazing work! Except of this change :)

**Comment by CEbbinghaus** [comment]
This topic has been argued about a lot but I wanted to voice my opinion on the matter since this is a tool that I am passionately using and the only thing I dislike about it is the fact that it doesn't have more widespread adoption. 

Removing the implicit `pre[scriptname]` hooks is an excellent idea since the original npm behaviour is extremely wrong. "preserve" is the example brought up the most but there are so many different reasons why this is bad ([see Fortran's implicit none](http://www.personal.psu.edu/jhm/f90/statements/implicit.html)). 

However, outright removing them and adding the command line argument I think is a bad solution too. The issue isn't their existence but their implicitness. The command-line argument should certainly be kept for backwards compatibility (to further grow pnpm as a drop-in for npm).

But I agree with @stalniy and @jmcdo29 that the right approach is to simply remove the implicit nature by explicitly saying `pre:[packagename]`. This removes the possibility of any conflicts with the English language and clearly delimits the `pre` and `post` scripts as being in their own folder (if thinking of the colons like slashes in a path). It would also clearly state that they are "reserved" script namespaces for the lifecycle events.

I love all that the PNPM team have done and I really hope to see this added. Since currently our two options are "use a third party script runner and define the behaviour yourself" and "use a configuration option that enables the bad implicit behaviour"

**Comment by zkochan** [comment]
Adding `pre:` and `post:` sounds fine.

**Comment by shellscape** [comment]
I would like to know that it would be a breaking change for anyone that had scripts named with those prefixes or other purposes. so that should probably be an opt-in

**Comment by CEbbinghaus** [comment]
> I would like to know that it would be a breaking change for anyone that had scripts named with those prefixes or other purposes. so that should probably be an opt-in

Knowing how many people's environments would break would indeed be an extremely valuable metric to decide the number of projects that would break because of this change.

However, I believe that the removal of the implicit hooks has had a significantly greater impact than the addition of `pre:` and `post:` prefixes could have since those would have to have been implemented manually by the developer (and would likely have had the same logic as what is being proposed)

**Comment by stabai** [comment]
> no, the standard lifecycle scripts like postinstall or prepublishOnly are not affected.

It would be nice if the documentation covered exactly which scripts are standard lifecycle scripts that are not affected. I happened to care about postinstall, so my question is answered from this thread, but it's not super clear from the documentation that even this is true: https://pnpm.io/cli/run#differences-with-npm-run

**Comment by slavafomin** [comment]
Ouch! This has seriously exploded me in the face today after migrating from npm to pnpm. I was expecting pnpm to be compatible with npm...

**Comment by beenotung** [comment]
@slavafomin you can use pnpm for installation and still use npm for running scripts

**Comment by maczyt** [comment]
configure `enable-pre-post-scripts=true` in `.npmrc` is work for me

**Comment by basememara** [comment]
Ouch, surprised by this too since impression was it has parity with `npm`.

Added this to the docs since migrations could be severely messed up: https://github.com/pnpm/pnpm.io/pull/495

**Comment by beenotung** [comment]
I think it will make less surprises if pnpm switch the configuration from `enable-pre-post-script` to `disable-pre-post-script`.

**Comment by shellscape** [comment]
agreed. every single one of my repos has that turned back on.

**Comment by tombohub** [comment]
GTP:

> Mental models are the frameworks within which we understand and navigate the world, built from our cumulative experiences and expectations. They simplify the complex reality into understandable concepts, guiding our interactions and decision-making processes. However, when these models are broken or challenged—say, by encountering something that defies our expectations, like a square circle—our ability to predict outcomes falters, leading to confusion, a reassessment of our understanding, or even a complete overhaul of our mental framework to accommodate the new reality.

**Comment by zkochan** [comment]
It will be reverted in v9

https://github.com/pnpm/pnpm/pull/7634
